src/crypto-providers/signingValidation.ts

```diff
@@ -0,0 +1,256 @@
+  }
+}
+
+const _countWitnessableItems = (unsignedTxParsed: _UnsignedTxParsed) => {
+  // we count stake registrations separately because they don't necessarily require a staking witness
+  let numStakeRegistrationItems = 0
+  let numStakeOtherItems = unsignedTxParsed.withdrawals.length
+  let numPoolColdItems = 0
+  unsignedTxParsed.certificates.forEach((cert) => {
+    switch (cert.type) {
+      case TxCertificateKeys.STAKING_KEY_REGISTRATION:
+        numStakeRegistrationItems += 1
+        break
+
+      case TxCertificateKeys.STAKING_KEY_DEREGISTRATION:
+      case TxCertificateKeys.DELEGATION:
+        numStakeOtherItems += 1
+        break
+
+      case TxCertificateKeys.STAKEPOOL_RETIREMENT:
+        numPoolColdItems += 1
+        break
+
+      default:
+        break
+    }
+  })
+  return { numStakeRegistrationItems, numStakeOtherItems, numPoolColdItems }
+}
+
+const validateOrdinaryTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateStakeCredentialsAreAllOfType(
+    unsignedTxParsed, StakeCredentialType.ADDR_KEY_HASH, Errors.ScriptStakeCredentialInOrdinaryTx,
+  )
+}
+
+const validateOrdinaryWitnesses = (params: SigningParameters) => {
+  const {
+    stakeSigningFiles, poolColdSigningFiles, mintSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  const {
+    numStakeRegistrationItems, numStakeOtherItems, numPoolColdItems,
+  } = _countWitnessableItems(params.unsignedTxParsed)
+
+  if (numStakeRegistrationItems + numStakeOtherItems === 0 && stakeSigningFiles.length > 0) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (numPoolColdItems === 0 && poolColdSigningFiles.length > 0) {
+    throw Error(Errors.TooManyPoolColdSigningFilesError)
+  }
+  if (!params.unsignedTxParsed.mint && mintSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validateOrdinarySigning = (params: SigningParameters) => {
+  validateOrdinaryWitnesses(params)
+  // these checks should be performed only with createSignedTx call (when the signing file set is
+  // expected to be complete) on top of validateOrdinaryWitnesses
+
+  const {
+    paymentSigningFiles, stakeSigningFiles, poolColdSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (paymentSigningFiles.length === 0) {
+    throw Error(Errors.MissingPaymentSigningFileError)
+  }
+
+  const {
+    numStakeOtherItems, numPoolColdItems,
+  } = _countWitnessableItems(params.unsignedTxParsed)
+
+  if (numStakeOtherItems > 0 && (stakeSigningFiles.length === 0)) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (numPoolColdItems > 0 && (poolColdSigningFiles.length === 0)) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
+}
+
+const validatePoolRegistrationTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateStakeCredentialsAreAllOfType(
```

Comment:
I guess we can drop this, it's not even clear what it does --- withdrawals are forbidden for pool registrations and there is a single certificate (though many pool owners, but we don't check them).

Reply:
Resolved in a thread above.